### PR TITLE
PIM-4749: Add transformation cookbook

### PR DESCRIPTION
**Pull request by willy-ahva**

### CODE REVIEW

File: cookbook/product_asset/add_new_transformation.rst [hunk 1]
```diff
@@ -0,0 +1,243 @@
+How to Add a New Transformation
+===============================
+
+Quick Overview
+--------------
+
+Transformations are classes which allow to transform a media from state A to state B.
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: media -> file

File: cookbook/product_asset/add_new_transformation.rst [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+For example, in the Product Asset specific case, they transform a Reference (A) to Variations (B).
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: a file coming from a reference to a variation file

File: cookbook/product_asset/add_new_transformation.rst [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+Each a Variation is a transformation of a Reference.
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: drop a

File: cookbook/product_asset/add_new_transformation.rst [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+Php5-imagemagick is used in Akeneo PIM but you can add you own library to transform your media.
+
+As Transformations are not specific PIM elements, they are stocked in `\\Akeneo` and not `\\PimEnterprise`.
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: located

File: cookbook/product_asset/add_new_transformation.rst [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+Then, you can find in ``Akeneo\Component\FileTransformer`` all business code and in ``Akeneo\Bundle\FileTransformerBundle`` all Symfony2 specific implementation.
+
+As Transformation run with the registry pattern, you just have to create your own Transformation implementing ``Akeneo\Component\FileTransformer\Transformation\TransformationInterface`` and to register the service with the ``akeneo_file_transformer.transformation`` tag.
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: drop the just

Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: and to register -> and register

File: cookbook/product_asset/add_new_transformation.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        public function __construct(
+            TransformationOptionsResolverInterface $optionsResolver,
+            ImageMagickLauncher $launcher,
+            array $supportedMimeTypes = ['image/jpeg', 'image/tiff', 'image/png']
+        ) {
+            $this->optionsResolver    = $optionsResolver;
+            $this->supportedMimeTypes = $supportedMimeTypes;
+            $this->launcher           = $launcher;
+        }
+
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: blank line

File: cookbook/product_asset/add_new_transformation.rst [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            $options  = $this->optionsResolver->resolve($options);
+
+            $fontPart = '-font ' . $options['font'];
+            $sizePart = '-pointsize ' . $options['size'];
+
+            $blackWatermarkPart = 'fill black text ' . $options['position'] . ' \'' . $options['watermark'] . '\'';
+            $whiteWatermarkPart = 'fill white text ' . $options['position'] . ' \'' . $options['watermark'] . '\'';
+            $drawPart = '-draw "gravity south ' . $blackWatermarkPart . ' ' . $whiteWatermarkPart . '"';
+
+            $cmd      = $fontPart . ' ' . $sizePart . ' ' . $drawPart;
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: please remove blank line or don't align

File: cookbook/product_asset/add_new_transformation.rst [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+            return $options;
+        }
+    }
+
+.. _OptionsResolver: http://symfony.com/doc/current/components/options_resolver.html
+.. note::
+    You can know more about this Symfony2 component on the OptionsResolver_ documentation.
+
+Next step is to add your services in existing files or in your own one :
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: remove space before :

Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: this sentence is weird

File: cookbook/product_asset/add_new_transformation.rst [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+.. code-block:: yaml
+
+    # src/Akeneo/Bundle/FileTransformerBundle/Resources/config/options_resolvers.yml
+    parameters:
+        [... existing parameters ...]
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: you can drop that

File: cookbook/product_asset/add_new_transformation.rst [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        akeneo_file_transformer.options.image.watermark.class: Akeneo\Component\FileTransformer\Options\Image\WatermarkOptionsResolver
+
+    services:
+        [... existing services ...]
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: same here

File: cookbook/product_asset/add_new_transformation.rst [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+.. _OptionsResolver: http://symfony.com/doc/current/components/options_resolver.html
+.. note::
+    You can know more about this Symfony2 component on the OptionsResolver_ documentation.
+
+Next step is to add your services in existing files or in your own one :
+
+.. code-block:: yaml
+
+    # src/Akeneo/Bundle/FileTransformerBundle/Resources/config/options_resolvers.yml
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: rename it to services.yml

File: cookbook/product_asset/add_new_transformation.rst [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    parameters:
+        [... existing parameters ...]
+        akeneo_file_transformer.options.image.watermark.class: Akeneo\Component\FileTransformer\Options\Image\WatermarkOptionsResolver
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: please use a simpler namespace beginning by Acme\CustomBundle

File: cookbook/product_asset/add_new_transformation.rst [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    services:
+        [... existing services ...]
+        akeneo_file_transformer.options.image.watermark:
+            class: %akeneo_file_transformer.options.image.watermark.class%
+
+.. code-block:: yaml
+
+    # src/Akeneo/Bundle/FileTransformerBundle/Resources/config/transformations.yml
+    parameters:
+        [... existing parameters ...]
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: samr

File: cookbook/product_asset/add_new_transformation.rst [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+    namespace Akeneo\Component\FileTransformer\Options\Image;
+
+    use Akeneo\Component\FileTransformer\Exception\InvalidOptionsTransformationException;
+    use Akeneo\Component\FileTransformer\Options\TransformationOptionsResolverInterface;
+    use Symfony\Component\OptionsResolver\OptionsResolver;
+
+    /**
+     * Watermark Options Resolver
+     */
+    class WatermarkOptionsResolver implements TransformationOptionsResolverInterface
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: I don't think that we need to use a dedicated option resolver for integrators

File: cookbook/product_asset/add_new_transformation.rst [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+            arguments:
+                - '@akeneo_file_transformer.options.image.watermark'
+                - '@akeneo_file_transformer.transformation.image.image_magick_launcher'
+            tags:
+                - { name: akeneo_file_transformer.transformation, alias: transformation_watermark }
+
+Translate Asset Transformation Details for Channels
+---------------------------------------------------
+
+A translation key is automatically created with the ``Watermark->getName()``. You can translate it in your own file or in ``src/PimEnterprise/Bundle/EnrichBundle/Resources/translations/messages.en.yml`` :
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: same here for the namespace

File: cookbook/product_asset/add_new_transformation.rst [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+.. code-block:: yaml
+
+    # src/PimEnterprise/Bundle/EnrichBundle/Resources/translations/messages.en.yml
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: namespace

File: cookbook/product_asset/add_new_transformation.rst [hunk 17]
```diff
@@ -0,0 +1,243 @@
+How to Add a New Transformation
+===============================
+
+Quick Overview
```
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: please add a disclaimer to explain that it's a enterprise edition feature

File: cookbook/product_asset/add_new_transformation.rst (hunk 11 above)
Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: That's the real file. I can help users to see where are the others services and parameters. If you want I can't mention this file before and change the file name.

File: cookbook/product_asset/add_new_transformation.rst [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+
+            return $options;
+        }
+    }
+
+.. _OptionsResolver: http://symfony.com/doc/current/components/options_resolver.html
+.. note::
+    You can know more about this Symfony2 component on the OptionsResolver_ documentation.
+
+Next step is to add your services in existing files or in your own one:
```
Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: rework this sentence

File: cookbook/product_asset/add_new_transformation.rst (hunk 14 above)
Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: Now it's done, I think i can't be a bad idea to keep it.

File: cookbook/product_asset/add_new_transformation.rst (hunk 17 above)
Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: done

File: cookbook/product_asset/add_new_transformation.rst (hunk 6 above)
Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: done

File: cookbook/product_asset/add_new_transformation.rst [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+Transformations are classes which allow to transform a file from state A to state B.
+For example, in the Product Asset specific case, they transform a file coming from a Reference (A) to a Variation file (B).
+Each Variation is a transformation of a Reference.
+
+Php5-imagemagick is used in Akeneo PIM but you can add you own library to transform your file.
+
+As Transformations are not specific PIM elements, they are located in `\\Akeneo` and not `\\PimEnterprise`.
+Then, you can find in ``Akeneo\Component\FileTransformer`` all business code and in ``Akeneo\Bundle\FileTransformerBundle`` all Symfony2 specific implementation.
+
+Transformation run with the registry pattern. You have to create your own Transformation implementing ``Akeneo\Component\FileTransformer\Transformation\TransformationInterface`` and register the service with the ``akeneo_file_transformer.transformation`` tag.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: runs

File: cookbook/product_asset/add_new_transformation.rst [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+Then, compiler pass will inject your tagged Transformation in the registry which will make it available.
+
+.. _registry: http://martinfowler.com/eaaCatalog/registry.html
+.. note::
+    Know more about registry_ pattern.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: Learn more

File: cookbook/product_asset/add_new_transformation.rst [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+This cookbook assumes that you already created a new bundle to add your new Transformation. Its namespace will be `Acme\CustomBundle`.
+
+Create a New Transformation
+---------------------------
+
+For the need of the cookbook, we'll add a watermark transformation.
+
+.. _watermark: http://www.imagemagick.org/Usage/annotating/#wmark_text
+.. note::
+    To know more about this transformation you can read the ImageMagick watermark_ documentation.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: To learn more

File: cookbook/product_asset/add_new_transformation.rst [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+In order to create your transformation, you have to follow these rules :
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: rules:

File: cookbook/product_asset/add_new_transformation.rst [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+         */
+        public function getName()
+        {
+            return 'watermark';
+        }
+    }
+
+.. _ImageMagickLauncher: https://github.com/akeneo/pim-enterprise-dev/blob/master/src/Akeneo/Component/FileTransformer/Transformation/Image/ImageMagickLauncher.php
+.. note::
+    To see how the imagemagick command is launched, you can take a look to the ImageMagickLauncher_.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: take a look at the ...

File: cookbook/product_asset/add_new_transformation.rst [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+                throw InvalidOptionsTransformationException::general($e, 'watermark');
+            }
+
+            return $options;
+        }
+    }
+
+.. _OptionsResolver: http://symfony.com/doc/current/components/options_resolver.html
+.. note::
+    You can know more about this Symfony2 component on the OptionsResolver_ documentation.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: learn more

File: cookbook/product_asset/add_new_transformation.rst [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+
+.. code-block:: csv
+
+    channel;configuration
+    mobile;{"watermark":{"size":"15", "watermark":"Copyright Akeneo"}}
+
+.. note::
+    Configuration is a json format. All these options are required but as we pushed a default value we can put nothing.
+
+Once you created CSV file you can go in Akeneo PIM and then start import with asset channel configuration import in csv profile.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: go to Akeneo PIM... start importing

File: cookbook/product_asset/add_new_transformation.rst [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+Be careful, if you import only the previous file, all your previous configuration will be removed. You need to add your own configuration in the file to keep it.
```
Comment (MarieMinasyan) [inline cookbook/product_asset/add_new_transformation.rst]: configurations

File: cookbook/product_asset/add_new_transformation.rst [hunk 27]
```diff
@@ -0,0 +1,235 @@
+How to Add a New Transformation
+===============================
+
+Quick Overview
+--------------
+
+**This cookbook is about a feature present only in the Enterprise Edition.**
```
Comment (rybus) [inline cookbook/product_asset/add_new_transformation.rst]: only provided

File: cookbook/product_asset/add_new_transformation.rst (hunk 14 above)
Comment (sumbobyboys) [inline cookbook/product_asset/add_new_transformation.rst]: not agree. It adds confusion to the integrator IMHO

Comment (willy-ahva) [inline cookbook/product_asset/add_new_transformation.rst]: confusion could have come not to do the same than our code imho

**Comment by nidup** [comment]
GTM

**Comment by rybus** [comment]
GTM

**Pull request merged**